python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -153,6 +156,23 @@ def get_output_blocking(self) -> MaybeRefBundle:
             except IndexError:
                 time.sleep(0.01)
 
+    def outqueue_memory_usage(self) -> int:
+        """Return the object store memory of this operator's outqueue.
+
+        Note: Python's deque isn't truly thread-safe since it raises RuntimeError
+        if it detects concurrent iteration. Hence we don't use its iterator but
+        manually index into it.
+        """
+
+        object_store_memory = 0
+        for i in range(len(self.outqueue)):
+            try:
+                bundle = self.outqueue[i]
+                object_store_memory += bundle.size_bytes()
+            except IndexError:
+                break  # Concurrent pop from the outqueue by the consumer thread.
```

Comment:
Should we put a lock and implement a thread-safe one, if it's not truly safe? This seems to cause inaccurate accounting and then affect scheduling decisions?

Reply:
I don't think there is any risk here, since the worst you can do is mis-count a couple elements. A lock would be much more invasive.